core/src/main/java/smile/classification/FLD.java

```diff
@@ -273,32 +278,51 @@ public FLD(double[][] x, int[] y, int L, double tol) {
         }
 
         T.setSymmetric(true);
-        EVD eigen = T.eigen();
+        EVD eigen = T.eigen();  
         
         tol = tol * tol;
         double[] s = eigen.getEigenValues();
         for (int i = 0; i < s.length; i++) {
             if (s[i] < tol) {
                 throw new IllegalArgumentException("The covariance matrix is close to singular.");
             }
-
-            s[i] = 1.0 / s[i];
-        }
-
+        }     
+        //
+        // calculate the square root of diagonal inverse of eigenvalue matrix from SVD of covariance matrix
+        //
         DenseMatrix U = eigen.getEigenVectors();
-        DenseMatrix UB = U.atbmm(B);
-
+        DenseMatrix D = eigen.getD().inverse();
```

Comment:
This is slow and also not numerically stable.  D is a diagonal matrix, which may have zeros in the diagonal. But `inverse()` is a general method, suitable for dense matrix.

Reply:
This is from the reference paper if what I understand is correct.
![image](https://user-images.githubusercontent.com/1487283/41448130-7862a800-708b-11e8-9d91-d8f913b5c55c.png)